src/org/traccar/protocol/Tk103ProtocolDecoder.java

```diff
@@ -206,7 +353,7 @@ protected Object decode(
             sentence = sentence.substring(beginIndex + 1);
         }
 
-        if (channel != null) {
+        if (channel != null && sentence.length() >= 16) {
```

Comment:
Why are you adding this?

Reply:
to prevent this error/exception:

`2017-11-20 00:58:10 DEBUG: [BE5D7A93: 5002 < x.x.x.xx] STR: (864555555555,ZC03,191117,235808,$1 .Sensor sensitivity: 1
2 .Alert status: Off
3 .Check interval is set to 240 minute(s).
4 .Checkgsm interval is set to 60 minute(s).
5 .SOS SMS Alert: On
6 .SOS Call Alert: On
7 . Power: 94%$) HEX: xxxxxxx

2017-11-20 00:59:54  WARN: [BE5D7A93] error - String index out of range: 12 - StringIndexOutOfBoundsException (... < Tk103ProtocolDecoder:356 < ExtendedObjectDecoder:58 < ...)

2017-11-20 00:59:54  INFO: [BE5D7A93] disconnected
`
Some "Command Results" contains very "free/random" text, which can include frame end symbol ")", which result in several frames, even with very small sizes. They are ignored everywhere, except in this code.

Reply:
Can I see proper original HEX? I don't see why it would not work.

Reply:
I have modified the traccar to log plain text, not hex, to easy debug...

Device sends to traccar:
(864555555555,ZC03,191117,235808,$1 .Sensor sensitivity: 1
2 .Alert status: Off
3 .Check interval is set to 240 minute(s).
4 .Checkgsm interval is set to 60 minute(s).
5 .SOS SMS Alert: On
6 .SOS Call Alert: On
7 . Power: 94%$)
(with \r\n as "new lines" but this does not matter for this case).

As tk103 protocol has ")" symbol as frame end symbol, this one message from device will be actually separated to 3 messages and "decode" will be called 3 times like this:

a) decode("(864555555555,ZC03,191117,235808,$1 .Sensor sensitivity: 1
2 .Alert status: Off
3 .Check interval is set to 240 minute(s");

b) decode(".
4 .Checkgsm interval is set to 60 minute(s");  <----- here will be exception,  because sentence = "s" - only one symbol. It is like that, because inside decode() sentence.indexOf('(') will remove everything until and including "(".

c) decode(".
5 .SOS SMS Alert: On
6 .SOS Call Alert: On
7 . Power: 94%$").

Reply:
Then the problem is with frame decoder. It needs to be fixed instead of implementing some workaround in the protocol decoder.

Reply:
Yes, i am agree that one frame decoder can't support all devices... For this device, seems like proper decoder should not try to search frame end symbol between "$" symbols...

Reply:
No, that's not what I said. In this case it's definitely possible to have single frame decoder.

Reply:
yes it is possible, but for this device the only solution to properly decode all data as single frame is to not try to search frame end symbol ")" between "$" symbols. Here is a format or this command from documentation:
"return content" as we can see also can contain frame end symbol ")", but it should be ignored. Hope that they does not send "$" symbol inside return content :)

1.4 Return SMS (ZC03)
(013632651491,ZC03, 040613, 040137,$...$)
"(" stands for packet start
"013632651491" stands for tracker ID
" ZC03" stands for data command
"040613" stands for date, 04/June/2013
"040137" stands for time, 04:01:37, UTC time
“$...$” stands for return content
")" stands for packet end

Reply:
I was thinking about counting opening and closing brackets, but ignoring content between `$` should also work.